[hd, hdx] Move tracing macros in InsertTask<T>

nvidia-jomiller opened the pull request:
### Description of Change(s)

This PR fixes a very hard to identify issue with link time optimization (-flto) for `hdx/taskController.cpp` where `GetRenderIndex()->InsertTask<HdxTask>()` would fail with cryptic error messages for release builds of USD with address sanitizers.

Effectively, `GetRenderIndex()->InsertTask<HdxPickTask>()` is defined in `hdx/unitTestDelegate.cpp` and `hdx/taskController.cpp`. The linker will, understandably, optimize out the symbol(s) in `hdx/taskController.o` since `hdx/unitTestDelegate.o` defines it first. This causes a linker error since it's `GetRenderIndex()->InsertTask<HdxPickTask>()` has been instrumented with ASan inside of `taskController.cpp`

An example of the error for a release build with address sanitizers enabled:

```bash
`_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_11HdxPickTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__' referenced in section `.data.rel.local' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o: defined in discarded section `.rodata._ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_11HdxPickTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__[_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_11HdxPickTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE16TraceKeyData_596]' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o
`_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_13HdxShadowTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__' referenced in section `.data.rel.local' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o: defined in discarded section `.rodata._ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_13HdxShadowTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__[_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_13HdxShadowTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE16TraceKeyData_596]' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o
`_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_18HdxSimpleLightTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__' referenced in section `.data.rel.local' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o: defined in discarded section `.rodata._ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_18HdxSimpleLightTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__[_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_18HdxSimpleLightTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE16TraceKeyData_596]' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o
`_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_16HdxSelectionTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__' referenced in section `.data.rel.local' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o: defined in discarded section `.rodata._ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_16HdxSelectionTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__[_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_16HdxSelectionTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE16TraceKeyData_596]' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o
`_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_13HdxRenderTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__' referenced in section `.data.rel.local' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o: defined in discarded section `.rodata._ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_13HdxRenderTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE8__func__[_ZZN32pxrInternal_v0_24__pxrReserved__13HdRenderIndex10InsertTaskINS_13HdxRenderTaskEEEvPNS_15HdSceneDelegateERKNS_7SdfPathEE16TraceKeyData_596]' of pxr/imaging/hdx/CMakeFiles/hdx.dir/taskController.cpp.o
```

The root of the issue is caused by the [HD_TRACE_FUNCTION in renderIndex.h](https://github.com/PixarAnimationStudios/OpenUSD/blob/v24.03/pxr/imaging/hd/renderIndex.h#L596). With this macro disabled, a release build of USD with address sanitizers will not encounter this error.

To circumvent this, the `HD_TRACE_FUNCTION` and `HF_MALLOC_TAG_FUNCTION` macros have been relocated into `_TrackDelegateTask`. The `_TrackDelegateTask` function now accepts a task creation function to create the actual `HdTask `. This prevents the weird link-time optimization error shown above.

### Fixes Issue(s)
-

<!--
Please follow the Contributing and Building guidelines to run tests against your
change. Place an X in the box if tests are run and are all tests passing.
-->
- [x] I have verified that all unit tests pass with the proposed changes
<!-- 
Place an X in the box if you have submitted a signed Contributor License Agreement.
A signed CLA must be received before pull requests can be merged.
For instructions, see: http://openusd.org/release/contributing_to_usd.html
-->
- [x] I have submitted a signed Contributor License Agreement

nvidia-jomiller committed:
[hdx] Define HdxTask subclasses for unitTestDelegate

In order to prevent incorrect link time optimizations with taskController.cpp
and address sanitizer, unitTestDelegate subclasses the necessary HdxTasks

nvidia-jomiller commented on pxr/imaging/hdx/unitTestDelegate.cpp:
```diff
@@ -58,7 +58,49 @@ TF_DEFINE_PRIVATE_TOKENS(
+// classes the linker will incorrectly "optimize" calls to the expanded
+// HD_TRACE_FUNCTION() in GetRenderIndex()->InsertTask<HdxPickTask>() for
+// taskController.cpp.
+class Hdx_UnitTestPickTask : public HdxPickTask {
```
Another potential solution to this problem would be to change the linking order for `hdx` to have `taskController.o` be linked before `unitTestDelegate.o`. This would present two problems:
1. The `unitTestDelegate` class would need to be moved into the `PUBLIC_CLASSES` section of `pxr_library` since private classes are configured first.
2. The link time optimization would still happen but with `uniTestDelegate.o` instead of `taskController.o`. Since `unitTestDelegate` is a helper for unit tests, disabling address sanitizer instrumentation for the offending `Add*Task(HdSceneDelegate*, const SdfPath&)` would be OK.

jesschimein commented:
Filed as internal issue #USD-9436

:heavy_exclamation_mark: Please make sure that a signed CLA has been submitted!

jesschimein commented:
/AzurePipelines run

nvidia-jomiller committed:
[hd, hdx] Move tracing macros in InsertTask<T>

HD_TRACE_FUNCTION and HF_MALLOC_TAG_FUNCTION macros have been relocated into _TrackDelegateTask
The _TrackDelegateTask function now accepts a task creation function to create the actual HdTask
This prevents a weird link-time optimization error when building with address sanitizers

jesschimein commented:
Please disregard the comment about the CLA -- we have you on file, Joshua!

jesschimein commented:
/AzurePipelines run
